lib/net.c

```diff
@@ -822,18 +822,54 @@ static int
 init_port(struct gatekeeper_if *iface, uint16_t port_id,
 	uint8_t *pnum_succ_ports)
 {
-	int ret = rte_eth_dev_configure(port_id, iface->num_rx_queues,
+	struct rte_eth_dev_info dev_info;
+	uint64_t configured_rss_hf;
+	int ret;
+
+	memset(&dev_info, 0, sizeof(dev_info));
+	rte_eth_dev_info_get(port_id, &dev_info);
+
+	/* Only use RSS hash functions this device can handle. */
+	configured_rss_hf = gatekeeper_port_conf.rx_adv_conf.rss_conf.rss_hf;
+	gatekeeper_port_conf.rx_adv_conf.rss_conf.rss_hf &=
+		dev_info.flow_type_rss_offloads;
+
+	/*
+	 * Check whether this device supports the configured RSS hashes.
+	 *
+	 * It seems common that devices do not exactly support the
+	 * hashes in the DPDK macros such as ETH_RSS_IP, so until we choose
+	 * set of minimum hash functions required (instead of ETH_RSS_IP
+	 * which is overkill), issue a warning in that case.
+	 *
+	 * TODO Find the minimum set of hash functions (ETH_RSS_*) that
+	 * Gatekeeper needs and set
+	 * gatekeeper_port_conf.rx_adv_conf.rss_conf.rss_hf accordingly.
+	 * Then, change this warning to an error.
+	 */
+	if (configured_rss_hf !=
+			gatekeeper_port_conf.rx_adv_conf.rss_conf.rss_hf) {
+		RTE_LOG(WARNING, PORT,
+			"Port %hhu does not support configured hash functions\n",
+			port_id);
+	}
+
+	ret = rte_eth_dev_configure(port_id, iface->num_rx_queues,
 		iface->num_tx_queues, &gatekeeper_port_conf);
 	if (ret < 0) {
 		RTE_LOG(ERR, PORT,
 			"Failed to configure port %hhu (err=%d)!\n",
 			port_id, ret);
-		return ret;
+		goto out;
 	}
 	if (pnum_succ_ports != NULL)
 		(*pnum_succ_ports)++;
 
-	return 0;
+	ret = 0;
```

Comment:
I think we need to reset what is in `gatekeeper_port_conf` because each port may be using a different NIC with a different set of supported hash functions. So if `gatekeeper_port_conf` says that we want hash functions X, Y and Z, but port 0 only supports X and Y, then that port will be configured with X and Y. But when we try to configure port 1 it should see that we are requesting hash functions X, Y and Z.

Reply:
Thanks! That makes sense.